Extract option logic from Grover class

willkoehler opened the pull request:
@abrom  I'm pushing up a WIP to get your feedback on the general shape of the solution. If it looks ok, I'll refine and add some specs.

More specifically
- What do you think about deriving `OptionsBuilder` from `Hash`? Too "tricky"? I could also just return the `options` hash from an accessor method or something.
- Also not sure on the best structure for `OptionsFixer.new` i.e. does `OptionsFixer.new(options).run` look like a reasonable API?
- I'm also seeing several spec errors like `expected: ["165", "42", "42"] got: [165, 42, 42]` on my system (/spec/grover_spec.rb:424) which is making it hard to know if I broke anything. These specs are failing even on the `Release v0.10.1` commit so I don't think they're caused by my changes. But I don't remember seeing these failures before on my system so that's kind of 🤔

willkoehler commented on lib/grover/options_builder.rb:
```diff
@@ -0,0 +1,43 @@
+      options = Utils.deep_stringify_keys(options)
+      Utils.deep_merge!(options, grover_configuration)
+      Utils.deep_merge!(options, meta_options) unless url_source?
+      self.merge!(OptionsFixer.new(options).run)
```
Good point. I realized the same thing about `self` this morning as I was waking up.
I agree `update` is clearer - much better than `merge`.

willkoehler commented:
> In any case, easy fix will be to modify the `mean_colour_statistics` method ~L578 in the `spec/grover_spec.rb` file to something like:
> 
> ```
>   def mean_colour_statistics(image)
>     %w[red green blue].map { |colour| image.data.dig('channelStatistics', colour, 'mean').to_s }
>   end
> ```

👍  I was thinking the same thing. Thanks for confirming. I'll add it to the PR.

willkoehler commented on lib/grover/options_builder.rb:
```diff
@@ -0,0 +1,43 @@
+    def initialize(options, url)
+      @url = url
+      options = Utils.deep_stringify_keys(options)
+      Utils.deep_merge!(options, grover_configuration)
```
Oh wow, nice catch. I was thinking of building onto the passed in options each step. But clearly that breaks the ability to override defaults. I'll fix it.

abrom replied:
if the specs didn't fail then it suggests I didn't have enough testing in there to validate it 😬

willkoehler replied:
I think the specs probably *were* failing. There were ~12 failures when I pushed the WIP.

abrom commented:
> ...and Code Climate is at it again 😬 https://codeclimate.com/github/Studiosity/grover/pull/46 But I was actually thinking of making a helper for all the fix_xx_options! methods that took an array of options and a transformation method/block. Maybe I'll look into that and see if that makes Code Climate happy.

I commented about that: https://github.com/Studiosity/grover/pull/46#discussion_r382964173

willkoehler commented on lib/grover/options_fixer.rb:
```diff
@@ -0,0 +1,55 @@
+    private
+
+    def fix_boolean_options!
+      %w[display_header_footer print_background landscape prefer_css_page_size].each do |opt|
```
I was thinking the same thing, but got lazy 😬 - and was worried it might be a little to "abstract". But your example reads really clearly. I'll put that in.

willkoehler commented:
> * That's odd.. must be something different about the `MiniMagick` install on your machine?!

It actually turned out to be a change in ImageMagick (I geeked out and dug through the source) https://github.com/ImageMagick/ImageMagick/commit/bc978e06ea3ae993b799e8c9a14ec84cdbfd3a4c#diff-b911bfc2e4a717df4fc996ef7decb2ffL713-R716

willkoehler commented:
Can you clarify what this does?
 https://github.com/Studiosity/grover/blob/d455f75d839dc47415aa2e35e6a754cc2031facf/lib/grover.rb#L322-L326

From what I can see `options['launch_args']` is usually (always?) an array.

Example: `["--disable-speech-api"]` in https://github.com/Studiosity/grover/blob/d455f75d839dc47415aa2e35e6a754cc2031facf/spec/grover_spec.rb#L275

So `return unless options['launch_args'].is_a? String` will return early and this method doesn't do anything?

willkoehler commented:
This spec is failing on my machine 

https://github.com/Studiosity/grover/blob/d455f75d839dc47415aa2e35e6a754cc2031facf/spec/grover_spec.rb#L115-L119

It's using the expectation in the first branch. But on my system, I have `MediaBox: [0, 0, 841.91998, 1189.91992]`, which matches the second expectation.

I'm running MacOS Catalina 10.15.3 with ImageMagick:
```
Version: ImageMagick 6.9.10-84 Q16 x86_64 2020-01-06 https://imagemagick.org
Copyright: © 1999-2020 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC Modules
Delegates (built-in): bzlib freetype jng jp2 jpeg lcms ltdl lzma png tiff webp xml zlib
```
I can ignore for now. May be due to the newer version of ImageMagick on my system?

willkoehler commented:
Also FYI this spec is failing.

https://github.com/Studiosity/grover/blob/d455f75d839dc47415aa2e35e6a754cc2031facf/spec/grover_spec.rb#L313-L317

With

```
expected Schmooze::JavaScript::Error with message matching /Failed to launch chrome! spawn \/totes\/invalid\/path/, got #<Schmooze::JavaScript::Error: Failed to launch the browser process! spawn /totes/invalid/path ENOENT
```

And this spec is failing

https://github.com/Studiosity/grover/blob/d455f75d839dc47415aa2e35e6a754cc2031facf/spec/grover_spec.rb#L395

With

```
Failure/Error: it { expect(image.data.dig('imageStatistics', 'all', 'mean').to_f).to be_within(1).of 97.7473 }
       expected 161.497 to be within 1 of 97.7473
```

These failures are on `master` and not caused by my changes. I can ignore for now.

willkoehler commented:
Please hold off on merge before I add a few specs...

abrom commented:
> Edit: Nevermind. I get it now. It converts strings that contains arrays into arrays.
> Ex `"['--something']"` --> `['--something']`

Correct. It allows the `launch_args` to be passed in through a meta tag (string value) and parse it as an array

abrom commented:
> Also FYI this spec is failing.
> 
> https://github.com/Studiosity/grover/blob/d455f75d839dc47415aa2e35e6a754cc2031facf/spec/grover_spec.rb#L313-L317
> 
> With
> 
> ```
> expected Schmooze::JavaScript::Error with message matching /Failed to launch chrome! spawn \/totes\/invalid\/path/, got #<Schmooze::JavaScript::Error: Failed to launch the browser process! spawn /totes/invalid/path ENOENT
> ```

Right. Puppeteer have been moving towards making the library driver agnostic and it looks like they've released a new version that does that (along with some changes to the error message copy). Best solution is likely going to be to make the copy match a regex to allow for `chrome` or `browser process`. You're welcome to make that change, or I'll get to it when I can.



> 
> And this spec is failing
> 
> https://github.com/Studiosity/grover/blob/d455f75d839dc47415aa2e35e6a754cc2031facf/spec/grover_spec.rb#L395
> 
> With
> 
> ```
> Failure/Error: it { expect(image.data.dig('imageStatistics', 'all', 'mean').to_f).to be_within(1).of 97.7473 }
>        expected 161.497 to be within 1 of 97.7473
> ```
> 
> These failures are on `master` and not caused by my changes. I can ignore for now.


I've never been a fan of the method used here, but it was a lot more reliable than pixel testing. It appears to be passing in Travis.. no doubt slightly different versions of the ImageMagick toolset etc

willkoehler commented:
> Right. Puppeteer have been moving towards making the library driver agnostic and it looks like they've released a new version that does that (along with some changes to the error message copy). Best solution is likely going to be to make the copy match a regex to allow for `chrome` or `browser process`. You're welcome to make that change, or I'll get to it when I can.

I added fix for this. See 0c9f50d90a611325de77148eec984c4f7be51d13

> I've never been a fan of the method used here, but it was a lot more reliable than pixel testing. It appears to be passing in Travis.. no doubt slightly different versions of the ImageMagick toolset etc

I added a variant to fix the spec on my system See 68593e5dd92d02c837753d7af5d2131ec34ed425. Feel free to pull if this isn't a direction you want to go.

willkoehler commented:
This is ready for another review. We might want to keep 561127f58e0e72312d1c598ff95e914a827703f2 un-squashed to document that change. I'm happy to rebase to keep the commits we want and squash everything else.

willkoehler commented:
Had some insight overnight. I'm adding some specs for `OptionsBuilder` please don't merge until those are in.

willkoehler commented:
`OptionsBuilder` specs are in. Everything is ready for a review 👍 

With the addition of `OptionsBuilder` specs, it looks like you can move some of the `Grover` specs to `OptionsBuilder` and eliminate a few of the dependencies on image content and PDF generation? I'm not sure of the full intention of those specs, so I'll leave that to you if it makes sense.

abrom commented on spec/grover/options_fixer_spec.rb:
```diff
@@ -0,0 +1,112 @@
+  context 'with an integer option' do
+    let(:options) { { 'viewport' => { 'width' => '123' } } }
+
+    it 'converts the option to a float' do
```
i think you meant to have `... to an integer` here ?

abrom commented on spec/grover/options_fixer_spec.rb:
```diff
@@ -0,0 +1,112 @@
+  context 'with a float option' do
+    let(:options) { { 'viewport' => { 'device_scale_factor' => '123.456' } } }
+
+    it 'converts the option to an integer' do
```
and `... to a float` here?

abrom commented on spec/grover/options_fixer_spec.rb:
```diff
@@ -0,0 +1,112 @@
+  context 'when a boolean option is truthy' do
+    let(:options) { { 'display_header_footer' => 'true' } }
+
+    it 'converts the options to a false literal' do
```
`... true ... `?

abrom commented on spec/grover/options_builder_spec.rb:
```diff
@@ -0,0 +1,69 @@
+    end
+    let(:options) { { wait_until: 'load' } }
+
+    it 'combines the meta options with the global config and passed-in options' do
```
it might be a good idea to update the options for each stage to show the full 'precedence' of the options. You're already testing that global settings are overridden by arguments above, so it should be enough to simply overload the argument settings with a meta setting?

willkoehler replied:
That makes sense. I'll add that.

abrom commented:
> This is ready for another review. We might want to keep [561127f](https://github.com/Studiosity/grover/commit/561127f58e0e72312d1c598ff95e914a827703f2) un-squashed to document that change. I'm happy to rebase to keep the commits we want and squash everything else.

I'm not too bothered either way. It's a pretty trivial fix and only relates to the specs so not sure it really needs much documenting

abrom commented:
> `OptionsBuilder` specs are in. Everything is ready for a review 👍
> 
> With the addition of `OptionsBuilder` specs, it looks like you can move some of the `Grover` specs to `OptionsBuilder` and eliminate a few of the dependencies on image content and PDF generation? I'm not sure of the full intention of those specs, so I'll leave that to you if it makes sense.

Not sure I following your meaning here? The image/PDF reader dependencies are necessary to validate the conversion process is doing what it's supposed to. How would this change eliminate that need?

willkoehler commented:
> Not sure I following your meaning here? The image/PDF reader dependencies are necessary to validate the conversion process is doing what it's supposed to. How would this change eliminate that need?

I though some of those specs may have just been checking that parameters were parsed and combined correctly. (Just a guess on my part since I don't have much context.) But if they are validating the conversion process, I agree, it wouldn't make sense to change them.

willkoehler commented:
> I'm not too bothered either way. It's a pretty trivial fix and only relates to the specs so not sure it really needs much documenting

Good point. We should just squash it in.

willkoehler commented on lib/grover/options_builder.rb:
```diff
@@ -0,0 +1,46 @@
+    def initialize(options, url)
+      @url = url
+      combined = grover_configuration
+      Utils.deep_merge!(combined, Utils.deep_stringify_keys(options))
```
LOL. I feel the opposite - parens make these lines easier to read 😬. (Probably due to my C++ roots.) But I'm happy to defer to your preference.

willkoehler commented on lib/grover/options_fixer.rb:
```diff
@@ -0,0 +1,58 @@
+
+    private
+
+    def fix_options!(*options)
```
Good point. What do you think about `paths`? Essentially `viewport.width` is a `.`-delimited path to the option.

willkoehler commented:
I fixed the last remaining broken spec on my system: c4abfb4efd1c342c0f69a1335416c738d954ba46

willkoehler commented:
The failure is a flickering spec: https://travis-ci.org/Studiosity/grover/jobs/655011626. Is there a way I can re-run these?

abrom commented:
No, but I can :)

I'm thinking I might add https://github.com/NoRedInk/rspec-retry at some point to help with the flake. It's often an issue with puppeteer just not booting. Bit odd..

willkoehler commented:
> Nice one @willkoehler, looks great! 👍

Awesome 🙌 Let me know if anything else comes up that I can help with. It's a pleasure working with you.

abrom commented:
Great, thanks @willkoehler. I can't think of any pressing changes needed at the moment. I'm currently updating the OptionsFixer to include some more of the missing type casting (some viewport options and some screenshot options).

There are plenty of things that Puppeteer can do that Grover doesn't support, but to be honest I'm not sure the merits of going all in on trying to get full API support.
